core/amber/src/main/scala/edu/uci/ics/texera/reporter/workFlowReporter.scala

```diff
@@ -0,0 +1,316 @@
+
+class workFlowReporter(
+                        @volatile private var configuration: Configuration,
+                        var relateActorPath: String,
+                        var relateWorkerIdentity: ActorVirtualIdentity
+) extends MetricReporter {
+  val expectedMetrics = configuration.metricList
+  import workFlowReporter._
+
+  private val logger = LoggerFactory.getLogger(classOf[workFlowReporter])
+  private val symbols = DecimalFormatSymbols.getInstance(Locale.US)
+  private val valueFormat = new DecimalFormat("#0.#########", symbols)
+  var actorMetrics = new mutable.HashMap[String, ListBuffer[MetricWrapper]]
+  symbols.setDecimalSeparator(
+    '.'
+  ) // Just in case there is some weird locale config we are not aware of.
+
+  override def stop(): Unit = {
+    logger.info(relateActorPath + " Stopped the workFlowWorker API reporter.")
+    val metricList = actorMetrics.getOrElse(relateActorPath, null)
+
+    if (metricList != null) {
+      logger.info(
+        s"""
+          | ActorId:$relateWorkerIdentity
+          | ActorPath:$relateActorPath
+          | Metric:
+          | ${metricList.mkString(
+          "-------------------",
+          "-------------------",
+          "\n-------------------"
+        )}
+          |""".stripMargin
+      )
+    }
+  }
+
+  override def reconfigure(config: Config): Unit = {
+    val newConfiguration = readConfiguration(config)
+    configuration = newConfiguration
+  }
+
+  override def reportPeriodSnapshot(snapshot: PeriodSnapshot): Unit = {
+    logger.info("reportPeriodSnapshot begin")
+    addData(snapshot)
+
+  }
+
+  private def addData(snapshot: PeriodSnapshot): Unit = {
+    val metricList = actorMetrics.getOrElse(relateActorPath, new ListBuffer[MetricWrapper])
+    val timestampStart = snapshot.from.toEpochMilli.toString
+    val timestampEnd = snapshot.to.toEpochMilli.toString
+    val interval = Math.round(Duration.between(snapshot.from, snapshot.to).toMillis())
+
+    def addDistribution(metric: MetricSnapshot.Distributions): Unit = {
+      var metricValueMap = new mutable.HashMap[String, String]
+      val unit = metric.settings.unit
+      metric.instruments.filter(p => {
+        pathFilter[Distribution](p)
+      }) foreach { d =>
+        val dist = d.value
+        val average = if (dist.count > 0L) (dist.sum / dist.count) else 0L
+        metricValueMap.put("avg", valueFormat.format(scale(average, unit)))
+        metricValueMap.put("count", valueFormat.format(dist.count))
+        metricValueMap.put("median", valueFormat.format(scale(dist.percentile(50d).value, unit)))
+        metricValueMap.put("max", valueFormat.format(scale(dist.max, unit)))
+        metricValueMap.put("min", valueFormat.format(scale(dist.min, unit)))
+        val metricUnit = unitToString(unit)
+        addMetric(
+          metric.name,
+          gauge,
+          d.tags,
+          metricValueMap,
+          metricUnit
+        )
+      }
+    }
+    def addMetric(
```

Comment:
Please add doc strings for all methods in this class, clearly define each methods usage, input, output, and if possible when they should be used.

Reply:
Especially for addMetric